### PR TITLE
Feature: cross check criteria constructor

### PR DESCRIPTION
**Pull request by andreikovna**
Add components for creating and downloading to backend Criteria Data for Cross-check.

### CODE REVIEW

File: client/src/modules/crossCheckCriteria/components/EditableTableForCrossCheck.tsx [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        }}
+        style={{wordBreak: 'break-word', fontStyle: 'normal'}}
+        size="small"
+        dataSource={dataCriteria}
+        columns={mergedColumns}
+        rowClassName="editable-row"
+        pagination={false}
+      />
+    </Form>
+  );
```
Comment (KostyaDedulya) [inline client/src/modules/crossCheckCriteria/components/EditableTableForCrossCheck.tsx:174]: two files

**Review by KostyaDedulya — CHANGES_REQUESTED** [review]

File: client/src/modules/crossCheckCriteria/components/EditableTableForCrossCheck.tsx (hunk 1 above)
Comment (andreikovna) [reply client/src/modules/crossCheckCriteria/components/EditableTableForCrossCheck.tsx]: Разбила на два компонента - таблица и ячейка

**Review by KostyaDedulya — APPROVED** [review]

File: client/src/components/Schedule/FormEntity.tsx [hunk 2]
```diff
@@ -65,24 +69,44 @@ const FormEntity: React.FC<Props> = ({
...
+    newDataCriteria.map((item, index) => (item.index = index));
+    setDataCriteria(newDataCriteria);
+  };
+
+  const addJSONtoCriteria = (criteria: string) => {
+    const oldCriteria = dataCriteria;
+    const addingCriteria = JSON.parse(criteria);
+    const newCriteria = [...oldCriteria, ...addingCriteria];
+    newCriteria.map((item, index) => (item.key = index.toString()));
+    newCriteria.map((item, index) => (item.index = index));
```
Comment (KostyaDedulya) [inline client/src/components/Schedule/FormEntity.tsx:94]: What is it?

File: client/src/components/Schedule/FormEntity.tsx [hunk 3]
```diff
@@ -65,24 +69,44 @@ const FormEntity: React.FC<Props> = ({
...
   const [isCrossCheckChecker, setIsCrossCheckChecker] = useState(checker);
   const isUpdateMode = editableRecord ? true : false;
 
+  const criteria = initialCriteria ? initialCriteria : [];
+  const [dataCriteria, setDataCriteria] = useState<CriteriaData[]>(criteria);
+
+  const addCriteria = (criteria: CriteriaData) => {
+    const newDataCriteria = dataCriteria.concat(criteria);
+    newDataCriteria.map((item, index) => (item.key = index.toString()));
+    newDataCriteria.map((item, index) => (item.index = index));
```
Comment (KostyaDedulya) [inline client/src/components/Schedule/FormEntity.tsx:85]: What is it?

**Review by KostyaDedulya — CHANGES_REQUESTED** [review]

File: client/src/components/Schedule/TableView.tsx [hunk 4]
```diff
@@ -258,12 +265,41 @@ export function TableView({
...
     }
 
     return [
+      {
+        title: 'Criteria For CrossCheck',
+        dataIndex: 'criteria',
+        render: (_: any, record: CourseEvent) => {
+          if (taskIdForCrossCheckCriteria.includes(record.event.id)) {
+            return (
+              <>
```
[the remark this answers is not in the document]
Comment (andreikovna) [reply client/src/components/Schedule/TableView.tsx]: removed it

File: client/src/modules/crossCheckCriteria/components/ModalCrossCheckCriteria.tsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    newDataCriteria.map((item, index) => (item.index = index));
+    setDataCriteria(newDataCriteria);
+  };
+
+  const addJSONtoCriteria = (criteria: string) => {
+    const oldCriteria = dataCriteria;
+    const addingCriteria = JSON.parse(criteria);
+    const newCriteria = [...oldCriteria, ...addingCriteria];
+    newCriteria.map((item, index) => (item.key = index.toString()));
+    newCriteria.map((item, index) => (item.index = index));
```
[the remark this answers is not in the document]
Comment (andreikovna) [reply client/src/modules/crossCheckCriteria/components/ModalCrossCheckCriteria.tsx]: add new function to avoid double code

**Review by KostyaDedulya — APPROVED** [review]

File: client/src/modules/crossCheckCriteria/components/ModalCrossCheckCriteria.tsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+export const ModalCrossCheckCriteria: React.FC<Props> = ({
+  isModalVisible,
+  setIsModalVisible,
+  dataCriteria,
+  setDataCriteria,
+  courseService,
+  taskId,
+}) => {
+  if (dataCriteria === undefined) {
```
Comment (igorlyatskiy) [inline client/src/modules/crossCheckCriteria/components/ModalCrossCheckCriteria.tsx:27]: Use strict comparisson

File: client/src/modules/crossCheckCriteria/components/ExportJSONButton.tsx [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+import _ from 'lodash';
+
+interface Props {
+  dataCriteria: CriteriaData[];
+}
+
+export function ExportJSONButton({ dataCriteria }: Props) {
+  const removeKeyAndIndex = (criteria: CriteriaData[]) => {
+    const newData = criteria.map(item => _.omit(item, ['key', 'index']));
+    return newData;
```
Comment (igorlyatskiy) [inline client/src/modules/crossCheckCriteria/components/ExportJSONButton.tsx:13]: `return criteria.map(item => _.omit(item, ['key', 'index'])); `

File: client/src/modules/crossCheckCriteria/components/ExportJSONButton.tsx [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  };
+
+  return (
+    <div style={{ textAlign: 'right' }}>
+      <Button icon={<FileOutlined />} style={{ marginTop: 15 }}>
+        <a
+          href={`data:text/json;charset=utf-8,${encodeURIComponent(JSON.stringify(removeKeyAndIndex(dataCriteria)))}`}
+          download="filename.json"
```
Comment (igorlyatskiy) [inline client/src/modules/crossCheckCriteria/components/ExportJSONButton.tsx:21]: crossCheckCriteria.json

**Review by igorlyatskiy — CHANGES_REQUESTED** [review]

File: client/src/modules/crossCheckCriteria/components/ExportJSONButton.tsx (hunk 7 above)
Comment (andreikovna) [reply client/src/modules/crossCheckCriteria/components/ExportJSONButton.tsx]: changed

**Review by igorlyatskiy — APPROVED** [review]

**Pull request merged**